outline.md

```diff
@@ -16,30 +61,127 @@
 
-* PATHs required by a given phase
-* Components used, probably by phase:
-  * commands? might be covered by paths
-  * interface: headers/modules/definitions that are needed at build but not runtime
-  * libraries: possibly also static vs dynamic
-* propagation: in A->B->C does A depend on C in a given phase
-* Does it have ABI impact?
+##### Build Plan
+* A *build plan* is an ordered list of *build phases*.
+
+### Environments
+*From [Environments](https://spack.readthedocs.io/en/latest/environments.html):*
+> An environment is used to group together a set of specs for the purpose of building, rebuilding and deploying in a coherent fashion.
+
+* A *spack environment*
+* What are they?
+  * How/Can these be serialized?
+  * How do these differ from (spack) modules?
+* Shell session?
+* Shared env vars with common parent process?
+
+## Package Developer Concepts
+
+### Package
+As per [Workflows](https://spack.readthedocs.io/en/latest/workflows.html#package-concrete-spec-installed-package), a *package* is a recipe to build one piece of software, containing:
+1. a mapping of version string to *package source*:
+  * **a package source will always resolve to a specific checksum**, but that checksum may change e.g. if it points to a git branch.
+  * a package source can be *unpacked* into a directory to modify the user's filesystem.
+1. a collection of specs representing the package's dependencies.
+  * these describe the dependencies for all versions of the package.
+    * e.g. these may contain *conditional dependencies*, which are [unified](#unified-constraints) during concretization.
+1. implementations for [build phases](#build-phases).
+1. any [environment modifications](#environment-modifications).
+
+These are always specified in a [`package.py`](#package-py) file in some [repo](#package-repositories).
+
+#### Environment Modifications
+* As per [unified environment](#unified-environment), the process environment can be treated as its own global *filesystem*.
 
+### Dependency Types
+
+Current dependency types:
+
+* Dependencies represent [implicit requirements](#implicit-requirements).
+* Currently, packages may *declare* either an [explicit](#explicit-dependencies) or [virtual](#virtual-dependendencies) dependency.
+  * These declared dependencies are intended to **satisfy implicit requirements**.
+
+#### Explicit Dependencies
+* An *explicit* dependency is a **human package developer's attempt** to satisfy any number of *implicit* requirements.
+  * *This description is based on [this model](https://github.com/spack/spack/discussions/20256#discussioncomment-248938).*
+* Currently, spack **conflates** *dependency types* with [build phases](#build-phases):
+  * The *dependency types* corresponding to each *build phase* can be viewed as **umbrellas** of implicit requirements.
+
+#### Implicit Requirements
+The known types of implicit requirements, sorted by the type of explicit dependency which is currently used to satisfy it:
```

Comment:
I think we should call these constraints or something -- not requierments.  I usually think of a "constraint" as something negative that restricts what you can do (which the stuff below mostly is).  A "requirement" is something positive -- some additional thing you need.  People also use "requirement" to mean "constraint" but I think we can be clearer.

Reply:
I agree, requirement feels very strict and also happens to be used in a specific way for python. Let's use constraints.